# RNN 情意分析 (使用IMDB 電影數據庫影評)

## 1. 初始準備

In [1]:
%matplotlib inline
import numpy as np
import matplotlib.pyplot as plt

In [2]:
%env KERAS_BACKEND = tensorflow

env: KERAS_BACKEND=tensorflow


In [4]:
from keras.models import Sequential
from keras.layers import Dense, Embedding, LSTM, Dropout
from keras.preprocessing import sequence

## 2. 讀入 IMDB 電影數據庫

In [5]:
from keras.datasets import imdb
(x_train, y_train), (x_test, y_test) = imdb.load_data(num_words=10000)

## 3. 整理資料格式

In [6]:
x_train = sequence.pad_sequences(x_train, maxlen=150)
x_test = sequence.pad_sequences(x_test, maxlen=150)

## 4. 決定 RNN 架構、訓練模型
### 4.1 model_1
* 文字壓成<span style="color:red;">5</span>維
* LSTM 有<span style="color:red;">4</span>個神經元

In [7]:
N = 5
K = 4

In [8]:
model_1 = Sequential()
model_1.add(Embedding(10000, N))
model_1.add(LSTM(K))
model_1.add(Dense(1, activation='sigmoid'))

Instructions for updating:
Colocations handled automatically by placer.


In [9]:
model_1.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_1.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, None, 5)           50000     
_________________________________________________________________
lstm_1 (LSTM)                (None, 4)                 160       
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 5         
Total params: 50,165
Trainable params: 50,165
Non-trainable params: 0
_________________________________________________________________


In [10]:
model_1.fit(x_train, y_train, batch_size=32, epochs=5, validation_data = (x_test, y_test))

Instructions for updating:
Use tf.cast instead.
Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 33s 1ms/step - loss: 0.5069 - acc: 0.7539 - val_loss: 0.3820 - val_acc: 0.8444
Epoch 2/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3003 - acc: 0.8848 - val_loss: 0.3626 - val_acc: 0.8458
Epoch 3/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.2260 - acc: 0.9192 - val_loss: 0.3459 - val_acc: 0.8590
Epoch 4/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.1873 - acc: 0.9352 - val_loss: 0.3522 - val_acc: 0.8590
Epoch 5/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.1530 - acc: 0.9496 - val_loss: 0.3801 - val_acc: 0.8555


### 4.2 model_2
* 文字壓成<span style="color:red;">1</span>維
* LSTM 有<span style="color:red;">4</span>個神經元

In [11]:
N = 1
K = 4

In [12]:
model_2 = Sequential()
model_2.add(Embedding(10000, N))
model_2.add(LSTM(K))
model_2.add(Dense(1, activation='sigmoid'))

In [13]:
model_2.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, None, 1)           10000     
_________________________________________________________________
lstm_2 (LSTM)                (None, 4)                 96        
_________________________________________________________________
dense_2 (Dense)              (None, 1)                 5         
Total params: 10,101
Trainable params: 10,101
Non-trainable params: 0
_________________________________________________________________


In [14]:
model_2.fit(x_train, y_train, batch_size=32, epochs=5, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.5372 - acc: 0.7291 - val_loss: 0.4050 - val_acc: 0.8394
Epoch 2/5
25000/25000 [==============================] - 31s 1ms/step - loss: 0.3120 - acc: 0.8838 - val_loss: 0.3412 - val_acc: 0.8604
Epoch 3/5
25000/25000 [==============================] - 31s 1ms/step - loss: 0.2427 - acc: 0.9117 - val_loss: 0.3243 - val_acc: 0.8664
Epoch 4/5
25000/25000 [==============================] - 31s 1ms/step - loss: 0.2021 - acc: 0.9292 - val_loss: 0.3338 - val_acc: 0.8627
Epoch 5/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.1743 - acc: 0.9413 - val_loss: 0.3742 - val_acc: 0.8587


### 4.3 model_3
* 文字壓成<span style="color:red;">3</span>維
* LSTM 有<span style="color:red;">50</span>個神經元

In [15]:
N = 3
K = 50

In [16]:
model_3 = Sequential()
model_3.add(Embedding(10000, N))
model_3.add(LSTM(K))
model_3.add(Dense(1, activation='sigmoid'))

In [17]:
model_3.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_3.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_3 (Embedding)      (None, None, 3)           30000     
_________________________________________________________________
lstm_3 (LSTM)                (None, 50)                10800     
_________________________________________________________________
dense_3 (Dense)              (None, 1)                 51        
Total params: 40,851
Trainable params: 40,851
Non-trainable params: 0
_________________________________________________________________


In [18]:
model_3.fit(x_train, y_train, batch_size=32, epochs=5, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 34s 1ms/step - loss: 0.4525 - acc: 0.7735 - val_loss: 0.3263 - val_acc: 0.8590
Epoch 2/5
25000/25000 [==============================] - 33s 1ms/step - loss: 0.2632 - acc: 0.8955 - val_loss: 0.3313 - val_acc: 0.8596lo
Epoch 3/5
25000/25000 [==============================] - 33s 1ms/step - loss: 0.2168 - acc: 0.9168 - val_loss: 0.3358 - val_acc: 0.8636
Epoch 4/5
25000/25000 [==============================] - 33s 1ms/step - loss: 0.1796 - acc: 0.9346 - val_loss: 0.3593 - val_acc: 0.8576
Epoch 5/5
25000/25000 [==============================] - 33s 1ms/step - loss: 0.1549 - acc: 0.9453 - val_loss: 0.4630 - val_acc: 0.8450


### 4.4 model_4
* 文字壓成<span style="color:red;">3</span>維
* LSTM 有<span style="color:red;">2</span>個神經元

In [19]:
N = 3
K = 2

In [20]:
model_4 = Sequential()
model_4.add(Embedding(10000, N))
model_4.add(LSTM(K))
model_4.add(Dense(1, activation='sigmoid'))

In [21]:
model_4.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_4.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_4 (Embedding)      (None, None, 3)           30000     
_________________________________________________________________
lstm_4 (LSTM)                (None, 2)                 48        
_________________________________________________________________
dense_4 (Dense)              (None, 1)                 3         
Total params: 30,051
Trainable params: 30,051
Non-trainable params: 0
_________________________________________________________________


In [22]:
model_4.fit(x_train, y_train, batch_size=32, epochs=5, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/5
25000/25000 [==============================] - 33s 1ms/step - loss: 0.5376 - acc: 0.7321 - val_loss: 0.4175 - val_acc: 0.8328
Epoch 2/5
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3420 - acc: 0.8713 - val_loss: 0.3707 - val_acc: 0.8453
Epoch 3/5
25000/25000 [==============================] - 31s 1ms/step - loss: 0.2787 - acc: 0.8962 - val_loss: 0.3737 - val_acc: 0.8489
Epoch 4/5
25000/25000 [==============================] - 31s 1ms/step - loss: 0.2380 - acc: 0.9155 - val_loss: 0.3809 - val_acc: 0.8472
Epoch 5/5
25000/25000 [==============================] - 31s 1ms/step - loss: 0.2081 - acc: 0.9289 - val_loss: 0.3829 - val_acc: 0.8449


### 4.5 model_5
* 文字壓成<span style="color:red;">3</span>維
* LSTM 有<span style="color:red;">4</span>個神經元
* activation function : <span style="color:red;">relu</span>
* 由於先前測試結果大約在epoch = 3 時的測試資料準確率最高，因此改為<span style="color:red;">epochs=3</span>

In [23]:
N = 3
K = 4

In [24]:
model_5 = Sequential()
model_5.add(Embedding(10000, N))
model_5.add(LSTM(K))
model_5.add(Dense(1, activation='relu'))

In [25]:
model_5.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_5.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_5 (Embedding)      (None, None, 3)           30000     
_________________________________________________________________
lstm_5 (LSTM)                (None, 4)                 128       
_________________________________________________________________
dense_5 (Dense)              (None, 1)                 5         
Total params: 30,133
Trainable params: 30,133
Non-trainable params: 0
_________________________________________________________________


In [26]:
model_5.fit(x_train, y_train, batch_size=32, epochs=3, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.5831 - acc: 0.7367 - val_loss: 0.3554 - val_acc: 0.8553
Epoch 2/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3064 - acc: 0.8869 - val_loss: 0.4114 - val_acc: 0.8618
Epoch 3/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3452 - acc: 0.8878 - val_loss: 0.4152 - val_acc: 0.8622


### 4.6 model_6
* 文字壓成<span style="color:red;">3</span>維
* LSTM 有<span style="color:red;">4</span>個神經元
* 使用Dropout <span style="color:red;">0.25</span>

In [27]:
N = 3
K = 4

In [28]:
model_6 = Sequential()
model_6.add(Embedding(10000, N))
model_6.add(Dropout(0.25))
model_6.add(LSTM(K))
model_6.add(Dense(1, activation='sigmoid'))

Instructions for updating:
Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.


In [29]:
model_6.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_6.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_6 (Embedding)      (None, None, 3)           30000     
_________________________________________________________________
dropout_1 (Dropout)          (None, None, 3)           0         
_________________________________________________________________
lstm_6 (LSTM)                (None, 4)                 128       
_________________________________________________________________
dense_6 (Dense)              (None, 1)                 5         
Total params: 30,133
Trainable params: 30,133
Non-trainable params: 0
_________________________________________________________________


In [30]:
model_6.fit(x_train, y_train, batch_size=32, epochs=3, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/3
25000/25000 [==============================] - 33s 1ms/step - loss: 0.5478 - acc: 0.7297 - val_loss: 0.4190 - val_acc: 0.8243
Epoch 2/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3641 - acc: 0.8555 - val_loss: 0.3647 - val_acc: 0.8468
Epoch 3/3
25000/25000 [==============================] - 33s 1ms/step - loss: 0.2951 - acc: 0.8878 - val_loss: 0.3601 - val_acc: 0.8482


### 4.7 model_7
* 文字壓成<span style="color:red;">3</span>維
* LSTM 有<span style="color:red;">4</span>個神經元
* 使用Dropout <span style="color:red;">0.4</span>

In [31]:
N = 3
K = 4

In [32]:
model_7 = Sequential()
model_7.add(Embedding(10000, N))
model_7.add(Dropout(0.4))
model_7.add(LSTM(K))
model_7.add(Dense(1, activation='sigmoid'))

In [33]:
model_7.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_7.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_7 (Embedding)      (None, None, 3)           30000     
_________________________________________________________________
dropout_2 (Dropout)          (None, None, 3)           0         
_________________________________________________________________
lstm_7 (LSTM)                (None, 4)                 128       
_________________________________________________________________
dense_7 (Dense)              (None, 1)                 5         
Total params: 30,133
Trainable params: 30,133
Non-trainable params: 0
_________________________________________________________________


In [34]:
model_7.fit(x_train, y_train, batch_size=32, epochs=3, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/3
25000/25000 [==============================] - 33s 1ms/step - loss: 0.5626 - acc: 0.7139 - val_loss: 0.4206 - val_acc: 0.8352
Epoch 2/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3698 - acc: 0.8504 - val_loss: 0.3583 - val_acc: 0.8452
Epoch 3/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.2890 - acc: 0.8844 - val_loss: 0.3195 - val_acc: 0.8680


### 4.8 model_8
* 文字壓成<span style="color:red;">3</span>維
* LSTM 有<span style="color:red;">4</span>個神經元
* 使用Dropout <span style="color:red;">0.5</span>

In [35]:
N = 3
K = 4

In [36]:
model_8 = Sequential()
model_8.add(Embedding(10000, N))
model_8.add(Dropout(0.5))
model_8.add(LSTM(K))
model_8.add(Dense(1, activation='sigmoid'))

In [37]:
model_8.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_8.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_8 (Embedding)      (None, None, 3)           30000     
_________________________________________________________________
dropout_3 (Dropout)          (None, None, 3)           0         
_________________________________________________________________
lstm_8 (LSTM)                (None, 4)                 128       
_________________________________________________________________
dense_8 (Dense)              (None, 1)                 5         
Total params: 30,133
Trainable params: 30,133
Non-trainable params: 0
_________________________________________________________________


In [38]:
model_8.fit(x_train, y_train, batch_size=32, epochs=3, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/3
25000/25000 [==============================] - 34s 1ms/step - loss: 0.5950 - acc: 0.6754 - val_loss: 0.4283 - val_acc: 0.8286
Epoch 2/3
25000/25000 [==============================] - 32s 1ms/step - loss: 0.3848 - acc: 0.8396 - val_loss: 0.3359 - val_acc: 0.8551
Epoch 3/3
25000/25000 [==============================] - 33s 1ms/step - loss: 0.3153 - acc: 0.8714 - val_loss: 0.3194 - val_acc: 0.8652


### 4.7 model_9
* 文字壓成<span style="color:red;">5</span>維
* LSTM 有<span style="color:red;">30</span>個神經元
* 使用Dropout <span style="color:red;">0.45</span>
* epochs = <span style="color:red;">2</span>

In [59]:
N = 5
K = 30

In [60]:
model_9 = Sequential()
model_9.add(Embedding(10000, N))
model_9.add(Dropout(0.4))
model_9.add(LSTM(K))
model_9.add(Dense(1, activation='sigmoid'))

In [61]:
model_9.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
model_9.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_14 (Embedding)     (None, None, 5)           50000     
_________________________________________________________________
dropout_9 (Dropout)          (None, None, 5)           0         
_________________________________________________________________
lstm_14 (LSTM)               (None, 30)                4320      
_________________________________________________________________
dense_14 (Dense)             (None, 1)                 31        
Total params: 54,351
Trainable params: 54,351
Non-trainable params: 0
_________________________________________________________________


In [62]:
model_9.fit(x_train, y_train, batch_size=32, epochs=2, validation_data = (x_test, y_test))

Train on 25000 samples, validate on 25000 samples
Epoch 1/2
25000/25000 [==============================] - 36s 1ms/step - loss: 0.4537 - acc: 0.7771 - val_loss: 0.3206 - val_acc: 0.8645
Epoch 2/2
25000/25000 [==============================] - 34s 1ms/step - loss: 0.2883 - acc: 0.8834 - val_loss: 0.3022 - val_acc: 0.8707


## 5. 結果

|模式|說明|epochs|訓練資料準確率|測試資料準確率
|---|---|---|---|---
|model|上課的範例|5|0.9459|0.8548
|model_1|增加文字壓的維度(3->5)|5|0.9496|0.8555
|model_2|減少文字壓的維度(3->1)|5|0.9413|0.8587
|model_3|增加 LSTM 神經元數量(4->50)|5|0.9453|0.8450
|model_4|減少 LSTM 神經元數量(4->2)|5|0.9289|0.8449
|model_5|activation function 改用 relu|3|0.8878|0.8622
|model_6|使用Dropout(0.25)|3|0.8878|0.8482
|model_7|使用Dropout(0.4)|3|0.8844|0.8680
|model_8|使用Dropout(0.5)|3|0.8714|0.8652
|model_9|文字壓成5維, LSTM 30個神經元, 使用Dropout(0.4)|2|0.8834|0.8707




## 6. 結論
1. 增加、減少文字壓到的維度、LSTM 神經元數量會使測試資料準確率些微上升
2. 減少文字壓到的維度、LSTM 神經元數量會使訓練資料準確率下降 
3. epochs = 2~3時測試資料準確率最高
4. 使用Dropout > 0.4 會使測試資料準確率上升

### 最後測試 model_9 得到 <span style="color:red;">87.07%</span> 的準確率